### DISCUSSION TITLE
inquiry about Retx of failed CSL Tx

### DISCUSSION THREAD

**Discussion by testter2** [openthread/openthread#9253] [Q&A] [answered]
Hi,

I have a question related to retransmission for csl packet in **SubMac::BeginTransmit**
as when **(error == kErrorInvalidState && mTransmitFrame.mInfo.mTxInfo.mTxDelay > 0)** then packet will be retransmitted using mTxDelay = 0 and mTxDelayBaseTime =0

what if first transmission time is near (less than device setup time) so transmission started before return of **Get<Radio>().Transmit(mTransmitFrame)**
so if this transmission failed due to any reason, then **Mac::HandleTransmitDone** will be called and **mOperation** will be **kOperationIdle** after it.

then for retransmission with (mTxDelay = 0, mTxDelayBaseTime =0) will assert inside **Mac::HandleTransmitDone** as the **mOperation**  in now **kOperationIdle**

Is this behavior expected ? why **SetState(kStateTransmit)** isn't called before retransmission?

Thanks in advance.

**Comment by Irving-cl** [comment, marked as answer]
Hi @testter2,

The openthread runs a tasklet loop: it runs tasks one by one.

The main loop of an OT app has something like this (`examples/apps/cli/main.c`):
```
    while (!otSysPseudoResetWasRequested())
    {
        otTaskletsProcess(instance);
        otSysProcessDrivers(instance);
    }
```

So back to the question, the entire function `SubMac::BeginTransmit` must have been completed before `Mac::HandleTransmitDone` is called. After `Radio::Transmit` is called, the platform API `otPlatRadioTransmit` is called and the radio platform will schedule a tx with the hardware. And action of radio platform (hardware) is another flow, the code will go on. Then if the radio platform want to notify that the tx has been done, it will not interrupt the main code flow and call the callback. Instead, it will do it in `otSysProcessDrviers`. Take nordic code as am example:
```
void otSysProcessDrivers(otInstance *aInstance)
{
    nrf5RadioProcess(aInstance);
    nrf5TransportProcess();
    nrf5TempProcess();
    nrf5AlarmProcess(aInstance);
}
```
The `TransmitDone` callback is called from `nrf5RadioProcess`. So even the actual radio tx is completed in short time, the callback will be called after the entire `BeginTransmit`.

**Reply by testter2** [reply]
Thanks for the detailed answer.
